### PR TITLE
Improve VA reservation robustness

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Bug fix** (*non-breaking change which fixes an issue*)
**New feature** (*non-breaking change which adds functionality*)

## Description:
On some systems, the amount of virtual address space is limited. This PR improves VA reservation rebustness by two means:
* retrying VA reservation with smaller size, if the default fails.
* trying to free unused VA if reducing the requested size doesn't work.

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: include/dali/core/mm/cuda_vm_resource.h [hunk 1]
```diff
@@ -191,13 +205,124 @@ class cuda_vm_resource : public memory_resource<memory_kind::device>,
...
+              for (int b = avail; b < end_avail; b++) {
+                blocks_to_free.push_back(region.unmap_block(b));
+                stat_.total_unmaps++;
+              }
+              auto *start_ptr = region.block_ptr<char>(avail);
+              auto *end_ptr = region.block_ptr<char>(end_avail);
+              free_mapped_.get_specific_block(start_ptr, end_ptr);
+
+              if (end_avail == end_block)
+                break;
```
Comment (JanuszL) [inline include/dali/core/mm/cuda_vm_resource.h:270]: Wouldn't this be covered by L259?

File: include/dali/core/mm/cuda_vm_resource.h [hunk 2]
```diff
@@ -191,13 +205,124 @@ class cuda_vm_resource : public memory_resource<memory_kind::device>,
...
+
+        auto *start_ptr = region.block_ptr<char>(rd.start_block);
+        auto *end_ptr = region.block_ptr<char>(rd.end_block);
+        ptrdiff_t range_size = (end_ptr - start_ptr);
+
+        if (region.empty()) {
+          if (tail.empty())
+            va_regions_.erase(va_regions_.begin() + rd.region_idx);
+          else
+            region = std::move(tail);
```
Comment (JanuszL) [inline include/dali/core/mm/cuda_vm_resource.h:304]: So in this case the `start_block` is the start of the old region?

File: include/dali/core/mm/cuda_vm_resource.h [hunk 3]
```diff
@@ -191,13 +205,124 @@ class cuda_vm_resource : public memory_resource<memory_kind::device>,
...
+        va_region tail = region.split(rd.end_block);
+        region.resize(rd.start_block);
+
+        auto *start_ptr = region.block_ptr<char>(rd.start_block);
+        auto *end_ptr = region.block_ptr<char>(rd.end_block);
+        ptrdiff_t range_size = (end_ptr - start_ptr);
+
+        if (region.empty()) {
+          if (tail.empty())
+            va_regions_.erase(va_regions_.begin() + rd.region_idx);
```
Comment (JanuszL) [inline include/dali/core/mm/cuda_vm_resource.h:302]: I don't fully understand that case - if both region and tail are empty does it mean that the old region was empty as well?

Comment (mzient) [reply include/dali/core/mm/cuda_vm_resource.h:302]: It means that the range covered the entire region: the tail is empty and the head is shrunk to zero blocks.

File: include/dali/core/mm/cuda_vm_resource.h (hunk 2 above)
Comment (mzient) [reply include/dali/core/mm/cuda_vm_resource.h:304]: Yes.

File: include/dali/core/mm/cuda_vm_resource.h (hunk 1 above)
Comment (mzient) [reply include/dali/core/mm/cuda_vm_resource.h:270]: If we already reach `end_block` there's no point in running the search in L271.

File: dali/core/mm/default_resource_test.cc [hunk 4]
```diff
@@ -379,7 +379,9 @@ static void ReleaseUnusedTestImpl(ssize_t max_alloc_size = std::numeric_limits<s
 }
 
 TEST(MMDefaultResource, ReleaseUnusedBasic) {
-  ReleaseUnusedTestImpl(256 << 20);
+  size_t free, total;
+  CUDA_CALL(cudaMemGetInfo(&free, &total));
+  ReleaseUnusedTestImpl(align_down(free / 2, 64 << 20));
```
Comment (mzient) [inline dali/core/mm/default_resource_test.cc:384]: This change was necessary because some systems (Orin) didn't even notice the 256M allocation.

**Review by JanuszL — APPROVED** [review]

**Review by banasraf — APPROVED** [review]

**Pull request merged**